File: 백준/Platinum/10217. KCM Travel/KCM Travel.cc
#include <iostream>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <cstring>

using namespace std;

#define endl '\n'
#define MAX 100+1
#define INF 987654321

int xd[4] = { 0,0,1,-1 };
int yd[4] = { 1,-1,0,0 };

int N, M, K;
vector<vector<pair<pair<int, int>, int>>> map;
vector<vector<int>> dist;

int dijkstra(int start) {
	priority_queue<pair<pair<int, int>, int>> pq;
	dist[start][0] = 0;
	pq.push({ {0,0},start });

	while (!pq.empty()) {
		int curDist = -pq.top().first.first;
		int curCost = -pq.top().first.second;
		int cur = pq.top().second;
		pq.pop();

		if (dist[cur][curCost] < curDist)
			continue;
		if (curCost > M)
			continue;
		if (cur == N)
			return curDist;

		for (int i = 0;i < map[cur].size();i++) {			
			int nextDist = map[cur][i].first.first + curDist;
			int nextCost = map[cur][i].first.second + curCost;
			int next = map[cur][i].second;

			if (nextCost > M)
				continue;
			if (nextDist < dist[next][nextCost]) {
				for (int j = nextCost + 1;j <= M;j++) {
					if (dist[next][j] <= nextDist)
						break;
					dist[next][j] = nextDist;
				}

				dist[next][nextCost] = nextDist;
				pq.push({ {-nextDist,-nextCost},next });
			}
		}
	}
	return -1;
}

int main(void)
{
	int Tc;
	cin >> Tc;
	while (Tc--) {
		cin >> N >> M >> K;
		map.assign(N + 1, vector<pair<pair<int, int>, int>>());
		dist.assign(N + 1, vector<int>(M + 1, INF));
		for (int i = 0;i < K;i++) {
			int u, v, c, d;
			cin >> u >> v >> c >> d;
			map[u].push_back({ {d,c},v });
		}
		int result = dijkstra(1);
		if (result == -1) {
			cout << "Poor KCM\n";
		}
		else {
			cout << result << '\n';
		}
	}
}

File: 백준/Platinum/10217. KCM Travel/KCM Travel.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    private static int N, M;
    private static List<Point>[] bridges;

    public static void main(String[] args) throws IOException {
        //입력
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        br.readLine();
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        bridges = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            bridges[i] = new ArrayList<>();
        }

        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            int d = Integer.parseInt(st.nextToken());

            bridges[u].add(new Point(v, c, d));
        }

        for (int i = 1; i <= N; i++) {
            bridges[i].sort((o1, o2) -> {
                return o1.time - o2.time;
            });
        }

        int answer = dijkstra();

        System.out.println(answer == Integer.MAX_VALUE ? "Poor KCM" : answer);
    }

    private static int dijkstra() {
        int answer = Integer.MAX_VALUE;

        int[][] dist = new int[N + 1][M + 1];
        for (int i = 1; i <= N; i++) {
            Arrays.fill(dist[i], Integer.MAX_VALUE);
        }
        dist[1][0] = 0;

        Queue<Point> q = new LinkedList<>();
        q.offer(new Point(1, 0, 0));
        while (!q.isEmpty()) {
            Point cur = q.poll();
            int curIdx = cur.idx;
            int curCost = cur.cost;
            int curTime = cur.time;

            if (curIdx == N) {
                answer = Math.min(answer, dist[curIdx][curCost]);
                continue;
            }

            if (answer < dist[curIdx][curCost]) {
                continue;
            }

            if (dist[curIdx][curCost] < curTime) {
                continue;
            }

            for (int i = 0; i < bridges[curIdx].size(); i++) {
                Point next = bridges[curIdx].get(i);
                int nextIdx = next.idx;
                int nextCost = next.cost + curCost;
                int nextTime = next.time + curTime;

                if (nextCost > M) {
                    continue;
                }

                if (dist[nextIdx][nextCost] <= nextTime) {
                    continue;
                }

                for (int j = nextCost; j <= M; j++) {
                    if (dist[nextIdx][j] > nextTime) {
                        dist[nextIdx][j] = nextTime;
                    } else {
                        break;
                    }
                }
                q.offer(new Point(nextIdx, nextCost, nextTime));
            }
        }

        for (int i = 0; i <= M; i++) {
            answer = Math.min(answer, dist[N][i]);
        }

        return answer;
    }

    static class Point {
        int idx, cost, time;

        public Point(int idx, int cost, int time) {
            this.idx = idx;
            this.cost = cost;
            this.time = time;
        }
    }
}